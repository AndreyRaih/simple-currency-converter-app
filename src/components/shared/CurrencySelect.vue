<template>
  <n-form-item :label-style="currencyFieldLabelStyle" :label="label">
    <n-select
      :placeholder="placeholder"
      :options="options"
      :value="value"
      :render-label="renderLabelWithSymbol"
      @update:value="(value) => $emit('update:value', value)"
    />
  </n-form-item>
</template>

<script lang="ts">
import { defineComponent, PropType, h } from 'vue'
import { NFormItem, NSelect, NText } from 'naive-ui'
import type { SelectOption } from 'naive-ui'
import { currencyFieldLabelStyle } from './constants'

export default defineComponent({
  name: 'CurrencySelect',
  components: {
    NFormItem,
    NSelect
  },
  props: {
    label: {
      type: String,
      required: true,
      default: null
    },
    placeholder: {
      type: String,
      required: true,
      default: null
    },
    options: {
      type: Array as PropType<SelectOption[]>,
      required: true,
      default: () => ([])
    },
    value: {
      type: String,
      default: null
    }
  },
  emits: [
    'update:value'
  ],
  setup () {
    const renderLabelWithSymbol = (option: SelectOption) => ([
      h(
        NText,
        {
          depth: 3,
          style: {
            marginRight: '8px'
          }
        },
        {
          default: () => option.value
        }
      ),
      option.label as string
    ])

    return {
      currencyFieldLabelStyle,
      renderLabelWithSymbol
    }
  }
})
</script>
