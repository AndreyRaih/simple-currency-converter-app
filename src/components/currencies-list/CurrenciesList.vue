<template>
  <currency-select
    label="Base currency"
    placeholder="Select base currency"
    :options="symbols"
    :value="baseCurrency"
    @update:value="(value) => $emit('base-currency-changed', value)"
  />
  <currencies-data-table
    :base-currency-code="baseCurrency"
    :loading="loading"
    :list="rates"
  />
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import CurrencySelect from '@/components/shared/CurrencySelect.vue'
import CurrenciesDataTable from './views/RatesTable.vue'
import { SelectOption } from 'naive-ui'
import { RatesDataTableItem } from '@/typings'

export default defineComponent({
  name: 'CurrenciesList',
  components: {
    CurrencySelect,
    CurrenciesDataTable
  },
  props: {
    baseCurrency: {
      required: true,
      type: String,
      default: null
    },
    symbols: {
      required: true,
      type: Array as PropType<SelectOption[]>,
      default: () => ([])
    },
    rates: {
      required: true,
      type: Array as PropType<RatesDataTableItem[]>,
      default: () => ([])
    },
    loading: {
      type: Boolean,
      default: false
    }
  },
  emits: [
    'base-currency-changed'
  ]
})
</script>
