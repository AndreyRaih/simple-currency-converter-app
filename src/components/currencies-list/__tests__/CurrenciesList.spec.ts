import { shallowMount } from '@vue/test-utils'
import CurrenciesList from '../CurrenciesList.vue'
import CurrencySelect from '@/components/shared/CurrencySelect.vue'

const defaultOptions = {
  props: {
    baseCurrency: 'USD',
    symbols: [],
    rates: [],
    loading: false
  }
}

describe('CurrenciesList.vue', () => {
  describe('Rendering', () => {
    it('should renders correctly', () => {
      const wrapper = shallowMount(CurrenciesList, defaultOptions)
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('Functional', () => {
    it('custom event @update:value should be handled correct', () => {
      const wrapper = shallowMount(CurrenciesList, defaultOptions)

      const handleSwitchCurrenciesTriggerElement = wrapper.findComponent(CurrencySelect)
      const updatedData = 'EUR';
      handleSwitchCurrenciesTriggerElement.vm.$emit('update:value', updatedData)

      const changeBaseCurrencyEvent = wrapper.emitted('base-currency-changed');
      expect(changeBaseCurrencyEvent).toHaveLength(1)
      expect((changeBaseCurrencyEvent as unknown[])[0]).toEqual([updatedData])
    })
  })
})
