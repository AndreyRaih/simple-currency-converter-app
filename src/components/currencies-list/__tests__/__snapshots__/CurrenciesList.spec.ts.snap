// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrenciesList.vue Rendering should renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <currency-select-stub
    label="Base currency"
    options=""
    placeholder="Select base currency"
    value="USD"
  />
  <currencies-data-table-stub
    basecurrencycode="USD"
    list=""
    loading="false"
  />
  
</div>
`;
